File: 2015/Day1/main.py
def calculate_final_floor(instructions):
    floor = 0
    for char in instructions:
        if char == '(':
            floor += 1
        elif char == ')':
            floor -= 1
    return floor

def find_basement_entry(instructions):
    floor = 0
    for position, char in enumerate(instructions, start=1):
        if char == '(':
            floor += 1
        elif char == ')':
            floor -= 1

        if floor == -1:
            return position
    
    return None

with open('input.txt', 'r') as file:
    instructions = file.read().strip()

print(f"Santa ends up on floor: {calculate_final_floor(instructions)}")
print(f"Santa enters the basement at position: {find_basement_entry(instructions)}")

File: 2015/Day1/main.dart
import 'dart:io';

int calculateFinalFloor(String instructions) {
  int floor = 0;
  for (var c in instructions.split('')) {
    if (c == '(')
      floor++;
    else if (c == ')') floor--;
  }
  return floor;
}

int findBasementEntry(String instructions) {
  int floor = 0;
  for (int position = 0; position < instructions.length; position++) {
    if (instructions[position] == '(')
      floor++;
    else if (instructions[position] == ')') floor--;

    if (floor == -1) return position + 1;
  }
  return -1;
}

void main() async {
  final file = File('input.txt');
  if (!await file.exists()) {
    print('File not found!');
    return;
  }

  final instructions = await file.readAsString();

  print('Part 1: ${calculateFinalFloor(instructions)}');
  print('Part 2: ${findBasementEntry(instructions)}');
}
